spi_dual_quad/pd.py

```diff
@@ -0,0 +1,356 @@
+
+        self.spibytes.append(Data(ss=ss, es=es, val=data))
+
+        # Bit annotations.
+        for bit in self.sio0bits:
+            self.put(bit[1], bit[2], self.out_ann, [ann_spi_sio0, ['%d' % bit[0]]])
+        for bit in self.sio1bits:
+            self.put(bit[1], bit[2], self.out_ann, [ann_spi_sio1, ['%d' % bit[0]]])
+
+        if self.is_quad:
+            for bit in self.sio2bits:
+                self.put(bit[1], bit[2], self.out_ann, [ann_spi_sio2, ['%d' % bit[0]]])
+            for bit in self.sio3bits:
+                self.put(bit[1], bit[2], self.out_ann, [ann_spi_sio3, ['%d' % bit[0]]])
+
+        # Dataword annotations.
+        self.put(ss, es, self.out_ann, [ann_spi_data, ['%02X' % self.spidata]])
+
+    def reset_decoder_state(self):
+        self.spidata = 0
+        self.sio0bits = []
+        self.sio1bits = []
+        self.sio2bits = []
+        self.sio3bits = []
+        self.bitcount = 0
+
+    def cs_asserted(self, cs):
+        active_low = (self.options['cs_polarity'] == 'active-low')
+        return (cs == 0) if active_low else (cs == 1)
+
+    def handle_bit(self, sio0, sio1, sio2, sio3, clk, cs):
+        # If this is the first bit of a dataword, save its sample number.
+        if self.bitcount == 0:
+            self.ss_block = self.samplenum
+            self.cs_was_deasserted = \
+                not self.cs_asserted(cs) if self.have_cs else False
+
+        ws = self.options['wordsize']
+        bo = self.options['bitorder']
+
+        if self.is_quad:
+            # Quad SPI SIO0 bits 4,0
+            # Quad SPI SIO1 bits 5,1
+            # Quad SPI SIO2 bits 6,2
+            # Quad SPI SIO3 bits 7,3
+            if bo == 'msb-first':
+                self.spidata |= sio3 << (ws - 1 - self.bitcount)
+                self.spidata |= sio2 << (ws - 1 - self.bitcount - 1)
+                self.spidata |= sio1 << (ws - 1 - self.bitcount - 2)
+                self.spidata |= sio0 << (ws - 1 - self.bitcount - 3)
+            else:
+                self.spidata |= sio3 << self.bitcount
+                self.spidata |= sio2 << (self.bitcount + 1)
+                self.spidata |= sio1 << (self.bitcount + 2)
+                self.spidata |= sio0 << (self.bitcount + 3)
+        else:
+            # Dual SPI SIO0 bits 6,4,2,0
+            # Dual SPI SIO1 bits 7,5,3,1
+            if bo == 'msb-first':
+                self.spidata |= sio1 << (ws - 1 - self.bitcount)
+                self.spidata |= sio0 << (ws - 1 - self.bitcount - 1)
+            else:
+                self.spidata |= sio1 << self.bitcount
+                self.spidata |= sio0 << (self.bitcount + 1)
+
+        # Guesstimate the endsample for this bit (can be overridden below).
+        es = self.samplenum
+        if self.bitcount > 0:
+            es += self.samplenum - self.sio0bits[0][1]
+
+        self.sio0bits.insert(0, [sio0, self.samplenum, es])
+        self.sio1bits.insert(0, [sio1, self.samplenum, es])
```

Comment:
Consider using a named tuple instead of a list for [bit,ss,es]

Reply:
It was based on spi code that use list, but I'll change for named tuple, the code will be easier to understand.

Reply:
I had forgotten that tuples are immutable. Yes, you could use constants to annotate. Something like `VAL,SS,ES = range(3)`. Alternatively, you could use a class instead of a named tuple. I will accept either option.

About the output compatibility, I understand that we need to keep the compatibility to allow the same stacked decoders. It does seem redundant, though, to include ES and SS in the data when they're already in the function parameters.